src/cargo/sources/registry/http_remote.rs

```diff
@@ -631,6 +706,116 @@ impl<'cfg> Downloads<'cfg> {
     }
 }
 
+mod www_authenticate {
+    use std::{collections::HashMap, mem::take};
+
+    enum State {
+        /// Either a challenge name or a parameter name.
+        NameOrKey,
+        /// A parameter value.
+        Value,
+        /// A quoted parameter value.
+        QuotedValue,
+    }
+
+    /// Parse a `www-authenticate` header into a map of "challenges" and their parameters.
+    /// See IETF RFC 7235 https://datatracker.ietf.org/doc/html/rfc7235#section-4.1 for details.
+    pub fn parse<T: AsRef<str>>(header_values: &[T]) -> HashMap<String, HashMap<String, String>> {
```

Comment:
I do not have a lot of confidence that this bespoke parser works correctly in all cases. Maybe it doesn't matter given that were only using it for a error message URL? (Although I would like to use it for the "challenge" in https://github.com/rust-lang/cargo/issues/10519) Is there some way to simplify it because we only care about data related to the `Cargo` name? Or use one of the parsing tools Cargo already depends on? Or is there a test suite of real `www-authenticate` headers we could test it on?

Reply:
Also [`<auth-scheme>`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate#directives) is case insensitive. So we need to handle that somewhere.

Reply:
Maybe use https://github.com/scottlamb/http-auth/blob/main/Cargo.toml with default=false, need to ask the team.

Reply:
Okay, I've switched to use the `http-auth` crate.